core/server/services/members/importer/index.js

```diff
@@ -4,50 +4,153 @@ const ObjectId = require('bson-objectid');
 const logging = require('../../../../shared/logging');
 
-const cleanupUndefined = (obj) => {
-    for (let key in obj) {
-        if (obj[key] === 'undefined') {
-            delete obj[key];
+const createInserter = table => data => insert(table, data);
+const createDeleter = table => data => del(table, data);
+
+const insertAllMembers = createInserter('members');
+const deleteMembers = createDeleter('members');
+const insertLabelAssociations = createInserter('members_labels');
+const insertCustomers = createInserter('members_stripe_customers');
+
+const doImport = async ({members, allLabelModels, importSetLabels, createdBy}) => {
+    const {
+        invalidMembers,
+        membersToInsert,
+        stripeCustomersToInsert,
+        labelAssociationsToInsert
+    } = getMemberData({members, allLabelModels, importSetLabels, createdBy});
+
+    const fetchedStripeCustomersPromise = fetchStripeCustomers(stripeCustomersToInsert);
+    const insertedMembersPromise = insertAllMembers(membersToInsert);
+
+    const insertedLabelsPromise = insertedMembersPromise
+        .then(() => insertLabelAssociations(labelAssociationsToInsert));
+
+    const insertedCustomersPromise = Promise.all([fetchedStripeCustomersPromise, insertedMembersPromise])
+        .then(function ([fetchedStripeCustomers]) {
+            return insertCustomers(fetchedStripeCustomers.customersToInsert);
+        });
+
+    const deletedMembersPromise = Promise.all([fetchedStripeCustomersPromise, insertedMembersPromise])
+        .then(function ([fetchedStripeCustomers]) {
+            return deleteMembers(fetchedStripeCustomers.membersToDelete);
+        });
+
+    // This looks sequential, but at the point insertedCustomersPromise has resolved so have all the others
+    const insertedCustomers = await insertedCustomersPromise;
+    const insertedMembers = await insertedMembersPromise;
+    const deletedMembers = await deletedMembersPromise;
+    const fetchedCustomers = await fetchedStripeCustomersPromise;
+    const insertedLabels = await insertedLabelsPromise;
+
+    const result = {
+        imported: {
+            count: insertedMembers.successful - deletedMembers.successful
+        },
+        invalid: {
+            count: insertedMembers.unsuccessful + deletedMembers.unsuccessful + invalidMembers.length,
+            errors: [
+                ...insertedMembers.errors,
+                ...insertedCustomers.errors,
+                ...insertedLabels.errors,
+                ...fetchedCustomers.errors
+            ]
         }
-    }
+    };
+
+    // Allow logging to happen outside of the request cycle
+    process.nextTick(() => {
+        // @TODO wrap errors with validation errors (or whatever is reasonable)
+        result.invalid.errors.forEach(err => logging.error(err));
+    });
+
+    return result;
 };
 
-function serializeMemberLabels(labels) {
-    if (_.isString(labels)) {
-        if (labels === '') {
-            return [];
+const CHUNK_SIZE = 5000;
```

Comment:
Would leave a note here about the CHUNK_SIZE value - we still need to test what's the optimal value for it is. Based on [these RoTs](http://mysql.rjweb.org/doc.php/ricksrots) the number should be between 100 and 1000.